pkg/didcomm/protocol/didexchange/states.go

```diff
@@ -391,63 +409,152 @@ func (ctx *context) handleInboundRequest(request *Request, options *options,
+	senderVerKey, err := recipientKey(responseDidDoc)
+	if err != nil {
+		return nil, nil, fmt.Errorf("handle inbound request: %w", err)
+	}
+
+	if ctx.doACAPyInterop {
+		// Interop: aca-py issue https://github.com/hyperledger/aries-cloudagent-python/issues/1048
+		responseDidDoc, err = convertPeerToSov(responseDidDoc)
+		if err != nil {
+			return nil, nil, fmt.Errorf("converting my did doc to a 'sov' doc for response message: %w", err)
+		}
+	}
+
+	response, err := ctx.prepareResponse(request, responseDidDoc, responseConnection)
+	if err != nil {
+		return nil, nil, fmt.Errorf("preparing response: %w", err)
+	}
+
+	connRec.TheirDID = requestConnection.DID
+	connRec.MyDID = responseConnection.DID
+	connRec.TheirLabel = request.Label
+
+	destination, err := service.CreateDestination(requestDidDoc)
 	if err != nil {
 		return nil, nil, err
 	}
 
+	if len(destination.MediaTypeProfiles) > 0 {
+		connRec.MediaTypeProfiles = destination.MediaTypeProfiles
+	}
+
+	// send exchange response
+	return func() error {
+		return ctx.outboundDispatcher.Send(response, senderVerKey, destination)
+	}, connRec, nil
+}
+
+func (ctx *context) prepareResponse(request *Request, responseDidDoc *did.Doc,
+	responseConnection *Connection) (*Response, error) {
 	// prepare the response
 	response := &Response{
 		Type: ResponseMsgType,
 		ID:   uuid.New().String(),
 		Thread: &decorator.Thread{
 			ID: request.ID,
 		},
-		ConnectionSignature: encodedConnectionSignature,
 	}
 
-	connRec.TheirDID = reqConn.DID
-	connRec.MyDID = connection.DID
-	connRec.TheirLabel = request.Label
+	if request.Thread != nil {
+		response.Thread.PID = request.Thread.PID
+	}
 
-	destination, err := service.CreateDestination(requestDidDoc)
+	// TODO followup: bring this out from under doACAPYInterop, and instead under an AIP 2.0 flag
+	//  part of issue https://github.com/hyperledger/aries-framework-go/issues/2495
+	if ctx.doACAPyInterop {
+		return ctx.prepareResponseWithSignedAttachment(request, response, responseDidDoc)
+	}
+
+	// prepare connection signature
+	encodedConnectionSignature, err := ctx.prepareConnectionSignature(responseConnection, request.Thread.PID)
 	if err != nil {
-		return nil, nil, err
+		return nil, fmt.Errorf("connection signature: %w", err)
 	}
 
-	if len(destination.MediaTypeProfiles) > 0 {
-		connRec.MediaTypeProfiles = destination.MediaTypeProfiles
+	response.ConnectionSignature = encodedConnectionSignature
+
+	return response, nil
+}
+
+func (ctx *context) prepareResponseWithSignedAttachment(request *Request, response *Response,
+	responseDidDoc *did.Doc) (*Response, error) {
+	if request.DocAttach != nil {
+		err := request.DocAttach.Data.Verify(ctx.crypto, ctx.kms)
+		if err != nil {
+			return nil, fmt.Errorf("verifying signature on doc~attach: %w", err)
+		}
 	}
 
-	senderVerKey, err := recipientKey(responseDidDoc)
+	var docBytes []byte
+
+	var err error
+
+	if ctx.doACAPyInterop {
```

Comment:
what's the acapy issue?

Reply:
https://github.com/hyperledger/aries-framework-go/pull/2733/files#diff-23f6a42334371adf3244d1e0517e14388a004b149c9dc05401b3c56abaea9368R156-R160